### ISSUE TITLE
Increment and decrement interpret text-num or num-num as a negative number

### ISSUE THREAD

**Issue by joshtriplett** [neovim/neovim#15562]
### Neovim version (nvim -v)

v0.4.4

### Vim (not Nvim) behaves the same?

yes, 8.2.2434

### Operating system/version

Linux, latest Debian

### Terminal name/version

GNOME Terminal 3.40.3 using VTE 0.64.2

### $TERM environment variable

xterm-256color

### Installation

Debian neovim package

### How to reproduce the issue

nvim -u NONE
Go into insert mode, and type `1.2.3-4`.
Leave insert mode, and put the cursor on the `4`.
Press C-a to increment.

### Expected behavior

`4` increments to `5. nvim should treat the `-` in `1.2.3-4` as a hyphen or other text punctuation, not a minus sign, because it's immediately preceded by an alphanumeric character. `x -4` (with a space) should increment to `x -3` treating the `-4` as a negative, but `x-4` should increment to `x-5` treating the `-` as a hyphen.

General rule: `-N` should not be treated as a negative number if `-` is immediately preceded by an alphanumeric character.

### Actual behavior

nvim treats the `-` in `1.2.3-4` as a minus sign, making the number `-4`, and "increments" it to `-3`.

**Comment by clason** [comment]
This is behavior inherited from vim; it would therefore be better to report it there so it gets fixed in both. (Neovim will port the vim patch once it is fixed.)

**Comment by chrisbra** [comment]
That sounds like adding some magic to the incrementing that may or may not work correctly depending on circumstances. Note: In newer Vims, you can simply set `:set nrformats+=unsigned` to circumvent that. Or you simply use visual mode and select exactly what you want to have incremented.

**Comment by clason** [comment]
(`nrformats+=unsigned` also works in current Neovim, by the way.)

**Comment by chrisbra** [comment]
I thought so, but my nvim.appimage 0.4.3 does not support it.

**Comment by clason** [comment]
Yes, but 0.4.4 is not the latest -- that is 0.5.0.

**Comment by clason** [comment]
I guess the main question here is whether Neovim should set `nrformats+=unsigned` by default according to the "principle of least surprise".

**Comment by joshtriplett** [comment]
`unsigned` forces all numbers to be treated as unsigned, but what I'm looking for is slightly different than that. I'd like a number with a minus sign to still be treated as a negative number, but only if the minus sign isn't immediately preceded by an alphanumeric character. So 4 in `x-4` or `12-4` should be treated as unsigned, but -4 in `x=-4` or `-3 -4` should be treated as negative.

**Comment by clason** [comment]
I understand, but let me be frank: we are _not_ going to add another option to a rarely used functionality that would lead to a lot of edge cases and make porting vim patches harder. (It's a reasonable request, granted, but so are many others that offer much broader benefit and core maintainer time is limited.) So unless either

1. vim changes the behavior in this situation (in which case we will of course port the patch) or
2. somebody makes a PR (that is sufficiently robust and contains enough tests),

the changed default option is the best you can hope for here.

**Comment by MuhammedZakir** [comment]
You may use a plugin to achieve this. See `nextval` plugin [1] for inspiration.

[1] https://github.com/qwertologe/nextval.vim/blob/master/plugin/nextval.vim

**Comment by chrisbra** [comment]
> unsigned forces all numbers to be treated as unsigned, but what I'm looking for is slightly different than that. 

Yes of course. That's why I mentioned visual mode as another alternative. It let's you select if you want to have the minus/hyphen included or not. 

I still think the smartinc behaviour would not gain anything, besides complaints about edge-cases from users, that do not understand the smart part :)

In any case, I guess you always know whether or not you want unsigned increment or not, so you could also map the increment to a function, that
a) either simply set's `unsigned` before incrementing (and afterwards resets it) or 
b) does the kind of smart detection of the chars in front of the character (If you go that way and find an algorithm that works reasonably well, let us know and we can think about implementing it in the core).

Thanks for understanding.

**Comment by clason** [comment]
So option 2 it is. I'll leave this open for the purpose of discussing the default change.

**Comment by justinmk** [comment]
> `unsigned` forces all numbers to be treated as unsigned, but what I'm looking for is slightly different than that.

`unsigned` is the least-surprising behavior. Only case (IME) where the signed behavior might be useful is for scripts. For humans, it is easy to just always regard the number as an absolute value.

**Comment by zeertzjq** [comment]
#29590 adds a new "blank" flag

**Comment by joshtriplett** [comment]
This has now been fixed in https://github.com/neovim/neovim/pull/29590 .
